### ISSUE TITLE
`no_std` without `alloc`

### ISSUE THREAD

**Issue by burdges** [ctz/rustls#283]
Just fyi, we're sorting out `#[no_std]` without `alloc` in https://github.com/RustCrypto/traits/pull/40 and https://github.com/SymbolicSoft/noiseexplorer/pull/34  I'm unsure if rustls maybe requires `std` for networking anyways, but anything we work out might prove handy internally regardless.

In fact, we could maybe define some trait suitable for rustls and noise, if not more generally, maybe something like:
```
pub trait AEAD {
    fn head_len(&self) -> usize;
    fn tail_len(&self) -> usize;
    fn overhead(&self) -> usize { self.head_len() + self.tail_len() }

    fn decrypt_detached(&mut self, head: &[u8], text: &mut [u8], tail: &[u8], ad: &[u8]);
    fn decrypt<'b>(&mut self, ciphertext: &'b mut [u8], ad: &[u8]) -> &'b mut [u8] {
        let (head,text) = ciphertext.split_at_mut(self.head_len());
        let (text,tail) = text.split_at_mut(ciphertext.len() - self.tail_len());
        self.decrypt_detached(head,text,tail,ad)
        text
    }

    fn encrypt_detached(&mut self, head: &mut [u8], text: &mut [u8], tail: &mut [u8], ad: &[u8]);
    fn encrypt<'b>(&'s mut self, buf: &'b mut [u8]) -> Encryptor<'b,'s,Self> {
        Encryptor { aead: self, buf, len: self.aead.head_len(), }
    }
}
pub struct Encryptor<'b,'s,T: AEAD> {
    buf: &'b mut [u8],
    aead: &'s mut T,
    len: usize,
}
impl<'b,'s,T: AEAD> Encryptor<'b,'s,T> {
    fn append(&mut self, len: usize) -> &mut [u8] {
        let old_len = self.len;
        self.len += len;
        self.buf.split_at_mut(old_len).1.split_at_mut(len).0
    }
    fn done(self, ad: &[u8]) -> &'b mut [u8] {
        let h = self.aead.head_len();
        let (head,more) = self.buf.split_at_mut(h);
        let (text,tail) = more.split_at_mut(self.len - h);
        let tail = tail.split_at_mut(self.aead.tail_len())
        self.aead.encrypt_detached(head,text,tail, ad);
    }
}
```

**Comment by MathiasKoch** [comment]
+1 on no_std support in general btw.

**Comment by burdges** [comment]
RustCrypto selected an AEAD trait emphasizes keeping prepended and appended information `Sized` using generic_array.  As a result, their AEAD trait is useless for any but the most narrow "use an AEAD here" applications.  Instead, we'd require some Handshake trait that resembled the above plus presumably stuff common to both TLS and Noise.  I'm unsure if such a trait actually makes sense for several reasons.  

You might still support preallocated buffers instead of alloc using inherent builder methods and helper struct like the above.

**Comment by DemiMarie-parity** [comment]
Does doing this correctly require const generics?

**Comment by burdges** [comment]
I do not think so..  RustCrypto surfaces all lengths with GenericArray, which should later become const generics.  I think here one should simply hide all the lengths, vaguely like I did above, but not really sure right now.

**Comment by brenzi** [comment]
why are you trying to avoid `alloc`? now that alloc is stable.
We would need the following function in `no_std` which need no networking anyway : 

```
let mut root_store = rustls::RootCertStore::empty();
    if root_store.add_pem_file(&mut ca_reader).is_err() {
        return Err("Failed to add CA");
    };
```
a partial support of `no_std`would be nice

**Comment by DemiMarie-parity** [comment]
@brenzi Some embedded platforms need to avoid all memory allocation, either because of real-time constraints or because they cannot afford the possibility that an allocation may fail.

**Comment by DemiMarie-parity** [comment]
@burdges what would the best way be to handle platforms that need all memory to be statically allocated?  Those are the platforms [BearSSL](https://www.bearssl.org) targets.

**Comment by burdges** [comment]
I donno currently.  In this issue, I wanted note that some `Handshake` trait(s) could cover both rustls and noise, while avoiding both std and alloc.  It does not sound like a high priority for anyone right now though, especially the part about working with both TLS and noise.  So, but  "because we can" to answer @brenzi  ;)

At present, Rust lacks ideal conventions for working with pre-allocated buffers.  I wrote down one convention with `Encryptor` above, but probably too specialized.  Instead, we should develop the `BufMut` trait idea from https://smallcultfollowing.com/babysteps/blog/2020/01/20/async-interview-5-steven-fackler/  

<small>In this vein, we could liberate the traits in `std::io` from `std` and `alloc` ala https://github.com/rust-lang/rust/issues/48331#issuecomment-570436716 but it require either a `SmallBox` type with alloc optional ala https://internals.rust-lang.org/t/idea-fixed-sized-traits/12179/8?u=jeffburdges or else an `io::Error` type that does trait objects manually, similar to `anyhow::Error` but with alloc optional.  I prefer the `SmallBox` with alloc optional approach myself, although it depends upon `&dyn Trait`, `&mut dyn Trait`, and `Box<dyn Trait>` having stable layout technically.

**Comment by DemiMarie-parity** [comment]
@burdges what about avoiding traits altogether?  A first step would be a completely non-generic version.

**Comment by briansmith** [comment]
> A first step could maybe be to allow no_std with an allocator ?

I have done this for webpki, culminating in https://github.com/briansmith/webpki/pull/175. The next version of webpki, by default, will target `no_std` environments without an allocator. The "alloc" feature of webpki will enable features that require liballoc (including its RSA support, for the time being). The "std" feature will further enable features of webpki that require libstd, which is basically `std::error` integration.

*ring* already does something similar.

Similarly, I submitted https://github.com/ctz/sct.rs/pull/6 to get sct.rs to support `no_std` environments that have liballoc.

I think the next step to make progress on this would be to do similar work for other dependencies of Rustls.

**Comment by briansmith** [comment]
I think support for custom allocators, along with other improvements for reducing/minimizing memory usage, would go a long way toward addressing most embedded needs, as the embedded application could pass in an arena allocator.

I'm a bit skeptical that there are a lot of people waiting to use Rustls in an environment where even such an arena allocator would be unacceptable. Even in an operating system kernel, there is a heap, in most cases.
